nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,215 @@
+#include <cmath>
+
+#include "geometry_msgs/msg/polygon_stamped.hpp"
+#include "tf2/transform_datatypes.h"
+
+#include "nav2_util/node_utils.hpp"
+#include "nav2_util/robot_utils.hpp"
+
+using rcl_interfaces::msg::ParameterType;
+
+namespace nav2_collision_monitor
+{
+
+PolygonSource::PolygonSource(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & source_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout,
+  const bool base_shift_correction)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout, base_shift_correction)
+{
+}
+
+PolygonSource::~PolygonSource()
+{
+  data_sub_.reset();
+}
+
+void PolygonSource::configure()
+{
+  Source::configure();
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<geometry_msgs::msg::PolygonStamped>(
+    source_topic, qos,
+    std::bind(&PolygonSource::dataCallback, this, std::placeholders::_1));
+}
+
+void PolygonSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data) const
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_.empty()) {
+    return;
+  }
+
+  tf2::Stamped<tf2::Transform> tf_transform;
+  for (const auto & polygon_stamped : data_) {
+    if (base_shift_correction_) {
+      // Obtaining the transform to get data from source frame and time where it was received
+      // to the base frame and current time
+      if (
+        !nav2_util::getTransform(
+          polygon_stamped.header.frame_id, polygon_stamped.header.stamp,
+          base_frame_id_, curr_time, global_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    } else {
+      // Obtaining the transform to get data from source frame to base frame without time shift
+      // considered. Less accurate but much more faster option not dependent on state estimation
+      // frames.
+      if (
+        !nav2_util::getTransform(
+          polygon_stamped.header.frame_id, base_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    tf2::doTransform(polygon_stamped, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+}
+
+void PolygonSource::convertPolygonStampedToPoints(
+  const geometry_msgs::msg::PolygonStamped & polygon, std::vector<Point> & data) const
+{
+  /* Generated by GPT */
```

Comment:
What part of it, specifically? Or the entire function?

Reply:
all of it (checked by me of course), edited